docs: hyperlinks: integration example (#6800)
wezterm/wezterm: docs/recipes/hyperlinks.md

## Requirements

By default, `ls` doesn't generate hyperlinks, here is an example of aliases to enable them in some common tools:

```sh
alias ls='ls --hyperlink --color=auto'
alias delta="delta --hyperlinks --hyperlinks-file-link-format='file://{path}#{line}'"
alias rg='rg --hyperlink-format=kitty'
```

> On macOS, you'll need to install `ls` from coreutils. For example using [homebrew](https://formulae.brew.sh/formula/coreutils),
> or use a modern alternative like [eza](https://github.com/eza-community/eza).

## Configuration

With this snippet, you can:

- **Click on an hyperlinked directory** to navigate into that directory and list its contents
- **Click on an hyperlinked file** and if its MIME type is 'text', open it directly in Neovim
- Other hyperlinks like URLs remain unchanged and follow WezTerm's default behavior

```lua
local wezterm = require("wezterm")
local act = wezterm.action
local config = wezterm.config_builder()

local function is_shell(foreground_process_name)
  local shell_names = { 'bash', 'zsh', 'fish', 'sh', 'ksh', 'dash' }

>   local process = string.gsub(foreground_process_name, '(.*[/\\])(.*)', '%2')

Reviewer: ```suggestion
  local process = string.match(foreground_process_name, "[^/\\]+$") or foreground_process_name
```

  for _, shell in ipairs(shell_names) do
    if process == shell then return true end
  end
  return false
end

wezterm.on('open-uri', function(window, pane, uri)
  local editor = 'nvim'

  if uri:find('^file:') == 1 and not pane:is_alt_screen_active() then
    -- We're processing an hyperlink and the uri format should be: file://[HOSTNAME]/PATH[#linenr]
    -- Also the pane is not in an alternate screen (an editor, less, etc)
    local url = wezterm.url.parse(uri)
    if is_shell(pane:get_foreground_process_name()) then
      -- A shell has been detected. Wezterm can check the file type directly
      -- figure out what kind of file we're dealing with
      local success, stdout, _ = wezterm.run_child_process({ 'file', '--brief', '--mime-type', url.file_path })
      if success then
        if stdout:find('directory') then
          pane:send_text(wezterm.shell_join_args({ 'cd', url.file_path }) .. '\r')

>           pane:send_text(wezterm.shell_join_args({ 'ls', '-a', '-p', '--group-directories-first' }) .. '\r')

Reviewer: ```suggestion
          pane:send_text(wezterm.shell_join_args({ 'ls', '-a', '-p', '--hyperlink', '--group-directories-first' }) .. '\r')
```

Author: The idea here is that the user will already have an alias for `ls` which is will be used in the local shell.

        .. '|| { test "$(file --brief --mime-type "$_f" | cut -d/ -f1 || true)" = "text" && '
        .. edit_cmd
        .. '; }; echo'
      pane:send_text(cmd .. '\r')
      return false
    end
  end

  -- without a return value, we allow default actions
end)

return config
```

## Optional configuration

By default, hyperlinks can be opened with a simple mouse click, without any modifier keys. If you find yourself accidentally clicking hyperlinks, you can customize this behavior.

Check the [mouse bindings](../config/mouse.md) section for examples on modifying the `OpenLinkAtMouseCursor` action.

For example, to require holding `CTRL` before opening a hyperlink, use the following configuration:

```lua
config.mouse_bindings = {
    {
      event = { Up = { streak = 1, button = "Left" } },
      mods = "CTRL",
      action = act.OpenLinkAtMouseCursor,
    },
    -- Disable the 'Down' event of CTRL-Click to avoid weird program behaviors
    {
      event = { Down = { streak = 1, button = "Left" } },
      mods = "CTRL",
      action = act.Nop,
    },
}
```

## Drawbacks

This setup sends the text of the commands directly into the active pane which has some drawbacks:

> - Does not work inside of any terminal program, the pane must be in a shell prompt

Reviewer: ```suggestion
- Does not work inside of any interactive terminal program that is not a shell; the pane must be in a shell prompt
```

- Requires an empty command prompt before clicking a hyperlink; otherwise, the command may not execute correctly

> - When a shell is not detected (see the `is_shell` function), WezTerm falls back to a long shell command, which may clutter the prompt slightly. This happens because there is no reliable way to determine a file's MIME type.

Reviewer: ```suggestion
- When a shell is not detected (see the `is_shell` function), WezTerm falls back to a long shell command, which may clutter the prompt slightly. This happens because WezTerm cannot directly determine a file's MIME type when not connected to a local shell.
```

# Use hyperlinks directly in the terminal

This recipe is an example on how to leverage hyperlinks directly in the terminal.
Hyperlinks are generated by terminal applications by using the [OSC-8 escape sequences](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda).

## Requirements

By default, `ls` doesn't generate hyperlinks, here is an example of aliases to enable them in some common tools:

```sh
alias ls='ls --hyperlink --color=auto'

> alias ll='ls --hyperlink -Ahl --group-directories-first --color=auto'

Reviewer 2: I think this page should focus on the hyperlink options in the aliases, other options like listing mode, color auto or grouping directories are a bit out of scope 🤔. 
Following this, I think the entire `ll` alias could be removed here.

What do you/others think?

Author: Totally true, I left the colors on `ls` tho

```lua
local wezterm = require("wezterm")
local act = wezterm.action
local config = wezterm.config_builder()

wezterm.on("open-uri", function(window, pane, uri)
  local editor = "nvim"

  if uri:find("^file:") == 1 then
    -- uri format is: file://[HOSTNAME]/PATH[#LINENR]
    local url = wezterm.url.parse(uri)

    -- figure out what kind of file we're dealing with
    local success, stdout, _ = wezterm.run_child_process({ "file", url.file_path })
    if success then
      if stdout:find("directory") then
        pane:send_text(
          wezterm.shell_join_args({ "cd", url.file_path }) .. "\r" .. wezterm.shell_join_args({ "ls", "-a", "-p", "--group-directories-first" }) .. "\r"
        )
        return false
      end

      if stdout:find("text") then
        if url.fragment then
          pane:send_text(wezterm.shell_join_args({ editor, "+" .. url.fragment, url.file_path }) .. "\r")
        else
          pane:send_text(wezterm.shell_join_args({ editor, url.file_path }) .. "\r")
        end
        return false
      end
    end
  end

  -- without a return value, we allow default actions
end)

config.mouse_bindings = {
  {

>     -- Trigger open selection (default is without a modifier)

Reviewer 2: Note that you should also disable Down event to avoid unexpected program behavior in some cases.
This is documented here:
https://wezterm.org/config/mouse.html#gotcha-on-binding-an-up-event-only

Author: You're right, without that config and tmux copy-mode is activated afterwards

    event = { Up = { streak = 1, button = "Left" } },
    mods = "CMD",
    action = act.OpenLinkAtMouseCursor,
  },
}

return config
```

## Drawback

> This setup requires that your command prompt be empty before clicking a hyperlink; otherwise, the sent command may not execute as expected.

Reviewer 2: Also note that this setup only works locally, not when we're connected on a remote machine

Reviewer 3: This does work over ssh, if the `file` check is bypassed (I checked). Since it utilizes just passing cmds to the pane, it should work regardless. so maybe also mentioning this would be helpful

# Use hyperlinks directly in the terminal

This recipe is an example on how to leverage hyperlinks directly in the terminal.
Hyperlinks are generated by terminal applications by using the [OSC-8 escape sequences](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda).

> 

Reviewer 2: I think you should add another paragraph describing what the experience of having this setup is.
For example mentioning that when you (ctrl/super-)click on a folder path in ls output, the shell will jump to that directory. Etc..

## Requirements

By default, `ls` doesn't generate hyperlinks, here is an example of aliases to enable them in some common tools:

```sh

> alias ls='ls --hyperlink --color=auto'

Reviewer 3: while this may beyond the initial scope of this recipe, it may be worthwhile to note that core `ls` is not the only `ls` tool which provides hyperlink support and could provide an alternative to the coreutils on MacOS if desired.

this is actually what I use and afaik maintains argument compatibility as well as additional theming and other features.

[eza-community/eza](https://github.com/eza-community/eza?tab=readme-ov-file)

  local editor = "nvim"

  if uri:find("^file:") == 1 then
    -- uri format is: file://[HOSTNAME]/PATH[#LINENR]
    local url = wezterm.url.parse(uri)

    -- figure out what kind of file we're dealing with
    local success, stdout, _ = wezterm.run_child_process({ "file", url.file_path })
    if success then
      if stdout:find("directory") then
        pane:send_text(
          wezterm.shell_join_args({ "cd", url.file_path }) .. "\r" .. wezterm.shell_join_args({ "ls", "-a", "-p", "--group-directories-first" }) .. "\r"
        )
        return false
      end

      if stdout:find("text") then
        if url.fragment then
          pane:send_text(wezterm.shell_join_args({ editor, "+" .. url.fragment, url.file_path }) .. "\r")
        else
          pane:send_text(wezterm.shell_join_args({ editor, url.file_path }) .. "\r")
        end
        return false
      end
    end
  end

  -- without a return value, we allow default actions
end)

config.mouse_bindings = {
    -- Trigger open selection (default is without a modifier)
    {
      event = { Up = { streak = 1, button = "Left" } },
      mods = "SUPER",
      action = act.OpenLinkAtMouseCursor,
    },

>     -- Disable the 'Down' event of CTRL-Click to avoid weird program behaviors

Reviewer 2: Note: The comment mentions CTRL, but the keybind uses SUPER..
What do you think about using CTRL instead for the keybind?

    {
      event = { Down = { streak = 1, button = "Left" } },
      mods = "SUPER",
      action = act.Nop,
    },
}

return config
```

## Drawbacks

> This setup works only locally (not over SSH) and requires an empty command prompt before clicking a hyperlink; otherwise, the command may not execute correctly.

Reviewer 2: It also requires that you are at a shell prompt!
If you're in neovim for example it'll give very unexpected behavior.

I'd suggest making a list of items for the limitations, it's easier to skim through visually without having to read a paragraph

```lua
local wezterm = require("wezterm")
local act = wezterm.action
local config = wezterm.config_builder()

wezterm.on("open-uri", function(window, pane, uri)
  local editor = "nvim"

  if uri:find("^file:") == 1 then
    -- uri format is: file://[HOSTNAME]/PATH[#LINENR]
    local url = wezterm.url.parse(uri)

    -- figure out what kind of file we're dealing with
    local success, stdout, _ = wezterm.run_child_process({ "file", "--brief", "--mime-type", url.file_path })
    if success then
      if stdout:find("directory") then
        pane:send_text(
          wezterm.shell_join_args({ "cd", url.file_path }) .. "\r" .. wezterm.shell_join_args({ "ls", "-a", "-p", "--group-directories-first" }) .. "\r"
        )
        return false
      end

      if stdout:find("text") then
        if url.fragment then
          pane:send_text(wezterm.shell_join_args({ editor, "+" .. url.fragment, url.file_path }) .. "\r")
        else
          pane:send_text(wezterm.shell_join_args({ editor, url.file_path }) .. "\r")
        end
        return false
      end
    end
  end

  -- without a return value, we allow default actions
end)

config.mouse_bindings = {

>     -- Trigger open selection (default is without a modifier)

Reviewer 4: I'm unclear why a ctrl-click binding is needed here; the text at the top of this example doesn't mention ctrl.
Should it?
If so, I think this comment should probably not mention the default behavior, as I think that makes this bit seem confusing to me: it implies that it is changing the default, but it isn't.

Author: Yes, you're right, I added that because without the modifier I was opening hyperlinks when I wanted to select text.
Maybe I should just mention that they can reconfigure the mouse bindings

    {
      event = { Up = { streak = 1, button = "Left" } },
      mods = "CTRL",
      action = act.OpenLinkAtMouseCursor,
    },
    -- Disable the 'Down' event of CTRL-Click to avoid weird program behaviors
    {
      event = { Down = { streak = 1, button = "Left" } },
      mods = "CTRL",
      action = act.Nop,
    },
}

return config
```

## Drawbacks

This setup sends the text of the commands directly into the active pane which has some drawbacks:

- Does not work inside of any terminal program, the pane must be in a shell prompt
- Requires an empty command prompt before clicking a hyperlink; otherwise, the command may not execute correctly

> - Only works locally (not over SSH, docker, etc)

Author: Great idea, this is what I came up with:

```lua
wezterm.on('open-uri', function(window, pane, uri)
  if uri:find('^file:') == 1 then
    -- uri format is: file://[HOSTNAME]/PATH[#linenr]
    local url = wezterm.url.parse(uri)

    local edit_cmd = url.fragment and 'nvim +' .. url.fragment .. ' "$_f"' or 'nvim "$_f"'
    local cmd = '_f="'
      .. url.file_path
      .. '"; { test -d "$_f" && { cd "$_f" ; ls -a -p --group-directories-first; }; } '
      .. '|| { test "$(file --brief --mime-type "$_f" | cut -d/ -f1 || true)" = "text" && '
      .. edit_cmd
      .. '; }; echo'

    pane:send_text(cmd .. '\r')

    return false
  end

  -- without a return value, we allow default actions
end)
```

I put an echo and the end to give some space after the cmd.
Regarding other filetypes, maybe it could fallback to `xdg-open` / `open`, or maybe that's overkill

> - Only works locally (not over SSH, docker, etc)

Reviewer 3: Although I believe trying to ensure all these different methods and edge cases is good, they may be nice for a more robust plugin possibly. 

Adding more validation and fallbacks I would think for a simple copy and paste recipe I think adds way more complexity than what I thought was a simple and elegant functionality IMO. 

A simple implementation that is well documented makes for a great starter for anyone that could then tailor this to their systems and use cases.

One thing you could do is utilize the `InputSelector` and give the user options as to how to handle the operations that may not be so straightforward.

> - Only works locally (not over SSH, docker, etc)

Author: I like the original implementation too just because it doesn't bloat the prompt/history as much. But it just breaks badly over ssh/docker and might cause damage if run inside a terminal app or editor.

I'll check if there's a reliable way to tell if the pane is running a shell/ssh/app to try to make it more robust and fallback to the long command only over ssh
